src/dir.targets

```diff
@@ -51,10 +56,25 @@
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets"
           Condition="'$(TargetFrameworkIdentifier)' != '.NETPortable'" />
 
+  <!-- Restore packages -->
+  <PropertyGroup>
+    <PackagesConfig Condition="Exists('$(MSBuildProjectDirectory)\packages.config')">$(MSBuildProjectDirectory)\packages.config</PackagesConfig>
+    <RestorePackages Condition="'$(RestorePackages)' == '' and Exists('$(PackagesConfig)')">true</RestorePackages>
+    <RestorePackagesSemaphore>$(PackagesDir)$(MSBuildProjectFile).package.restored</RestorePackagesSemaphore>
+  </PropertyGroup>
+  <Target Name="RestorePackages" 
+          BeforeTargets="ResolveNuGetPackages"
+          Inputs="$(PackagesConfig)"
+          Outputs="$(RestorePackagesSemaphore)"
+          Condition="'$(RestorePackages)' == 'true'">
+    <Exec Command="$(NugetRestoreCommand) &quot;$(PackagesConfig)&quot;" StandardOutputImportance="Low" />
+    <Touch Files="$(RestorePackagesSemaphore)" AlwaysCreate="true" />
```

Comment:
It is mainly for performance reasons. On my local machine doing an incremental build without this check almost doubles the time from 1.5 mins to 2.5 mins. Looking at my MSBuild performance logs with and with out this:

Without the semaphore file:
       2206 ms  ResolveAssemblyReferences                145 calls
       4705 ms  BuildPackages                              1 calls
     10111 ms  CopyTestToTestDirectory                   35 calls
     11639 ms  ResolveNuGetPackages                     136 calls
     16917 ms  **RestoreTestRuntimePackage**                 35 calls
     33685 ms  ResolveProjectReferences                 145 calls
     47918 ms  RunTestsForProject                        33 calls
     70324 ms  **RestorePackages**                          136 calls
    306540 ms  BuildAllProjects                           2 calls

With the semaphore file:
      2483 ms  ResolveAssemblyReferences                145 calls
      4687 ms  BuildPackages                              1 calls
     10200 ms  CopyTestToTestDirectory                   35 calls
     10709 ms  ResolveProjectReferences                 145 calls
     12349 ms  ResolveNuGetPackages                     136 calls
     45045 ms  RunTestsForProject                        33 calls
    158219 ms  BuildAllProjects                           2 calls

They aren't even in the picture. While I agree with you people could get into an invalidate state it should be easily fixable via a clean build and I would prefer people not pay the cost all the time for the 80-90% case.

From the logs you can see the next costly item which is the RunTestsForProject and CopyTestToTestDirectory which I believe you are going to be looking into and if you cannot I will get to them as some point as well.